### PR TITLE
Add NIX_AUTO_RUN/NIX_AUTO_INSTALL handling

### PR DESCRIPTION
**Pull request by matthewbauer**
This adds in the autorun/autoinstall handling for the bash scripts.

I've also fixed a change to the multiple args handling that broke ZSH scripts. The new fix should work in both Bash and ZSH.

Some other formatting changes as well.

### CODE REVIEW

File: command-not-found.sh [hunk 1]
```diff
@@ -4,36 +4,80 @@
...
+            # environment variables
+            # they are based on the ones in command-not-found.sh
+            #   NIX_AUTO_INSTALL : install the missing command into the
+            #                      user’s environment
+            #   NIX_AUTO_RUN     : run the command transparently inside of
+            #                      nix shell
+            # these will not return 127 if they worked correctly
+
+            # QUESTION: will this mess up some scripts if a broken script
+            #           fails to install?
```
Comment (bennofs) [inline command-not-found.sh]: Perhaps we should return `127` if the command failed to install?

Comment (matthewbauer) [inline command-not-found.sh]: That would work fine for ```nix-env -iA```, but I'm not sure how to handle it with nix-shell. How can you differentiate between nix-shell installing something and running a command that fails and nix-shell just failing? Maybe we could run nix-build ahead of time to ensure that it builds correctly (although that seems hacky).

File: command-not-found.sh [hunk 2]
```diff
@@ -4,36 +4,97 @@
...
+            # these will not return 127 if they worked correctly
+
+            if ! [ -z "$NIX_AUTO_INSTALL" ]; then
+                >&2 cat <<EOF
+The program '$cmd' is currently not installed. It is provided by
+the package '$toplevel.$attrs', which I will now install for you.
+EOF
+                nix-env -iA $toplevel.$attrs
+                if [ "$?" -eq 0 ]; then
+                    $@ # TODO: handle pipes correctly if AUTO_RUN/INSTALL is possible
```
Comment (bennofs) [inline command-not-found.sh]: Is this how `command-not-found` works in NixOS? From looking at the source at https://github.com/NixOS/nixpkgs/blob/a35ec5dda688f6eedbb281671ee11261c769610e/nixos/modules/programs/command-not-found/command-not-found.pl#L34, it seems it does not run the program, only installs it and exits with `126`.

Comment (matthewbauer) [inline command-not-found.sh]: It then runs it here though:

https://github.com/NixOS/nixpkgs/blob/a35ec5dda688f6eedbb281671ee11261c769610e/nixos/modules/programs/command-not-found/command-not-found.nix#L53

Comment (bennofs) [inline command-not-found.sh]: Oh, I missed that. Fine then

**Pull request merged**